File: tree/cpp/b15900.cpp
#include <iostream>
#include <vector>
#include <queue>

int N;
unsigned long long cnt = 0;
int visited[500001]{ 0 };

std::vector<std::vector<int>> adj;

void bfs(const int& start) {
    visited[start] = 1;
    std::queue<int> check;
    check.push(start);

    bool flag = true;
    while (check.size()) {
        int check_node = check.front();
        check.pop();
        flag = true;
        for (auto& adj_node : adj[check_node]) {
            if (visited[adj_node] == 0) {
                visited[adj_node] = visited[check_node] + 1;
                check.push(adj_node);
                flag = false;
            }
        }
        if (flag) {
            cnt += visited[check_node] - 1;
        }
    }

}

int main() {
    std::cin >> N;
    adj.resize(N + 1);
    int node1, node2;
    for (int i = 0; i < N - 1; ++i) {
        std::cin >> node1 >> node2;
        adj[node1].push_back(node2);
        adj[node2].push_back(node1);
    }

    bfs(1);
    if (cnt % 2) std::cout << "Yes" << '\n';
    else std::cout << "No" << '\n';
}

File: tree/python/b15900.py
from collections import deque
from sys import stdin
input = stdin.readline

def bfs(start):
    global cnt
    check = deque()
    check.append(start)
    visited[start] = 1
    flag = True
    while check:
        flag = True
        check_node = check.popleft()
        for adj_node in adj[check_node]:
            if visited[adj_node] == 0:
                visited[adj_node] = visited[check_node] + 1
                check.append(adj_node)
                flag = False
        if flag:
            cnt += visited[check_node] - 1

N = int(input())
adj = [[] for _ in range(N + 1)]
visited = [0] * (N + 1)
for i in range(N - 1):
    temp1, temp2 = map(int, input().split())
    adj[temp1].append(temp2)
    adj[temp2].append(temp1)
cnt = 0
bfs(1)
if cnt % 2:
    print("Yes")
else:
    print("No")
